BLSQ/iaso
# IA-2779: improve payment audit

Need audit trail for changes in Payments and Payment logs

Related JIRA tickets : IA-2779

## Self proofreading checklist


## Doc
- In the code

## Changes

### Backend
- Move `PaymentLot` creation and `mark_as_read`  update to `Tasks`, because these imply bulk operations and Audit logging requires a lot of DB read+write
- Add `payment_lot` foreign key to `PotentialPayment`, to allow filtering out `PotentialPayments` that are part of a `PaymentLot` that's being processed by a `Task`, and prevent assigning a `PotentialPayment` to several `PayementLot`s (and also filtering them out in the front-end view)
- Add `task` foreign key to `Payment Lot` and include it to the API payload to help the front-end handle `PaymentLot`s with an ongoing task in the UI
- Add Audit logging for `Payment Lot` creation:
    - Log  `Payment Lot` at creation, before the `Task` is launched: `name`, `comment` and `potential_payments` have already been defined
    - Log `Payment` when they are created within the task
    - Log each `OrgUnitChangeRequest` when the created `Payment` is created and the foreign key is set.
    - Log `PaymentLot` again when the task is done running and all expected values have been set
- Add Audit logging for "mark_as_sent" update:
    - Log each changed `Payment`
    - Log `PaymentLot` after the task has run, because the `status` has been updated.
- Add Audit logging when updating a single `Payment`:
    - For the `Payment`itself`
    - For the `PaymentLog` if its `status` has changed (eg: all payments are sent so the lot status is also "sent")
- Remove `AuditMixin` from payment related viewsets, because the logging is complex (nested) and required custom logic
- Add `AuditLogger` class to help with manually creating Audit logs
- use JSONWidget for `Modification` admin
- Define permissions to access logs for `Payment`, `PaymentLog` and `OrgUnitChangeRequest` in `/api/logs/{log_id}`
- Remove custom `save`method for `PaymentLot` model, as well as the "auto-update" of payment lot when a payment is updated
- Add `TaskAPITestCase` to help testing task related features

### Front-end

- Payment Lots page:
    - Show row with yellow background color, 0.5 opacity and all buttons disabled if the `PaymentLot`'s task is either `RUNNING` or `QUEUED`
- Payment Lot modal, if the `PaymentLot`'s task is either `RUNNING` or `QUEUED`:
    - Show payments rows with yellow background color, 0.5 opacity and all buttons disabled 
    - Set the `Table`  state to loading (ie: show spinner)
    - Disable Select All, Unselect All and Edit Selection buttons 
- Remove 'camelCase` warning from es-lint

## How to test

- Create change requests with "approved" status, preferably by several users, so there will be several potential payments
- start the worker!!
- Go to Potential Payments
- Create a Potential payment and save
- Go to Tasks: you should see the task
- Go to Payment Lots: the new payment lot should appear with the right data (number of changes, etc)
- Open the payment lot modal: play around with the selection (select all etc):
    - Try bulk saving with the "Edit Selected" button: this should launch a task
    - try single Payment saving with the Edit icon button 
- In the django admin, go to Audit > Modifications: you should see modifications created according the rules describes in the "changes section of this PR"

To test the layout changes (yellow rows)
- Either use asyncio to slow down the task execution in Python code
- Or change the display condition so the change is triggered when the task's status is `SUCCESS`

## Print screen / video

General flow

https://github.com/BLSQ/iaso/assets/38907762/2e6095a8-14fc-4c01-9415-e21b7e331b90

When task is running

![Screenshot 2024-03-21 at 15 14 14](https://github.com/BLSQ/iaso/assets/38907762/e329569a-e0c5-477f-b1a9-89940951bf74)

![Screenshot 2024-03-21 at 15 18 00](https://github.com/BLSQ/iaso/assets/38907762/8eda3f27-70bd-4830-9488-475aeffec360)

**UPDATE 22/03/2024: added refresh button**
![Uploading Screenshot 2024-03-22 at 17.07.51.png…]()



## Notes

Merges in https://github.com/BLSQ/iaso/pull/1192
